### PR TITLE
add encoders

### PR DESCRIPTION
**Pull request by Pnom**

### CODE REVIEW

File: src/routes/v1/utils/validateDecoderRequest.js [hunk 1]
```diff
@@ -10,7 +10,7 @@ const isRequestValid = ( {data, direction, hardwareType} ) => (typeof data === '
     && (!hardwareType || !hardwareTypes.has(hardwareType));
 
 
-export default ( {body}, reply, done ) => {
+export const validateDecoderRequest = ( {body}, reply, done ) => {
```
Comment (DarkPark) [inline src/routes/v1/utils/validateDecoderRequest.js]: how about `requestDecoder.js` and export function `validate`?

or even `request.js` and keep here everything related

File: package.json [hunk 2]
```diff
@@ -6,8 +6,9 @@
     "author": "",
     "scripts": {
         "lint": "eslint .",
-        "api:serve": "redoc-cli serve --watch api/openapi.json -p 3030 --ssr",
-        "api:build": "redoc-cli bundle api/openapi.json -o api/index.html",
+        "api:serve": "redocly serve --watch ./api/openapi.yaml -p 3030 --ssr",
+        "api:build": "redocly bundle ./api/openapi.yaml",
```
Comment (DarkPark) [inline package.json]: `        "api:bundle": "redocly bundle ./api/openapi.yaml",`
?

File: src/routes/v1/utils/validateGeneralRequest.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
+import {validateDecoderRequest, validateEncoderRequest} from './validateRequest.js';
```
Comment (DarkPark) [inline src/routes/v1/utils/validateGeneralRequest.js]: image

it should be ok to remove `validate` from each file name and use them for any related helpers

File: src/routes/v1/utils/validateMtxAesParameters.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
+import {mtx, utils} from 'jooby-codec/index.js';
+import errors from '../../../errors.js';
+
+
+const {accessLevels} = mtx.constants;
+const accessLevelsSet = new Set(Object.values(accessLevels));
+
+
+const validateAesParameters = ( request, reply ) => {
```
Comment (DarkPark) [inline src/routes/v1/utils/validateMtxAesParameters.js]: `export const validateMtxAesDecoderParameters = ( request, reply ) => {`

?

File: src/routes/v1/utils/validateMtxRequest.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+
+export const validateMtxDecoderRequest = ( request, reply ) => (
+    validateDecoderRequest(request, reply) || !validateMtxAesDecoderParameters(request, reply)
+);
+
+
+export const validateMtxEncoderRequest = ( request, reply ) => {
+    const {body} = request;
+    const {
```
Comment (DarkPark) [inline src/routes/v1/utils/validateMtxRequest.js]: in such simple cases it's better to put it in a single line

**Comment by DarkPark** [comment]
for a commit comment it's better to use `:` to separate topic from message, for example
instead of `controllers/decoders. add common utils` use `controllers/decoders: add common utils`

File: src/controllers/utils/prepareCommands.js [hunk 6]
```diff
@@ -1,7 +1,17 @@
-export const prepareCommand = command => {
-    const {parameters, message, constructor: {id, name}} = command;
+import {utils} from 'jooby-codec/index.js';
```
Comment (DarkPark) [inline src/controllers/utils/prepareCommands.js]: `src/controllers/utils/preparations.js` ? or something

as there is not only single prepareCommands inside

File: src/controllers/utils/prepareCommands.js [hunk 7]
```diff
@@ -1,7 +1,17 @@
...
 
-    return {id, name, parameters, message};
+
+export const prepareCommand = ( command, bytesConversionFormat ) => {
+    const json = command.toJson(bytesConversionFormat);
+    const {constructor: {id, name}} = command;
+
+    return json ? {id, name, ...JSON.parse(json)} : {id, name};
```
Comment (DarkPark) [inline src/controllers/utils/prepareCommands.js]: can `json` be an empty string? a `"null"` string? an empty object `"{}"` string?

maybe firstly create `let result = {id, name};` and then extend it with parsed json in try/catch without conditional checks?

File: src/controllers/decoders/mtxLora.js [hunk 8]
```diff
@@ -1,101 +1,67 @@
...
-export default async function decode ( request, reply ) {
-    const {message} = analog;
-    const {deviceEUI, data} = request.body;
-    const analogCommands = [];
-    let mtxMessage;
-
-    // eslint-disable-next-line import/namespace
-    const direction = directions ? directions[request.body.direction.toUpperCase()] : undefined;
+export default function decode ( {body}, reply ) {
+    const {
```
Comment (DarkPark) [inline src/controllers/decoders/mtxLora.js]: imo still can be in one line

File: src/controllers/decoders/obisObserver.js [hunk 9]
```diff
@@ -1,10 +1,49 @@
...
 
 /**
  * @this fastify.FastifyInstance
  */
-export default async function decode ( request, reply ) {
-    commonDecode(obisObserver, request, reply);
+export default function decode ( {body}, reply ) {
+    const {
+        bytes,
+        framingFormat,
```
Comment (DarkPark) [inline src/controllers/decoders/obisObserver.js]: not `frameFormat`?

File: src/controllers/decoders/obisObserver.js [hunk 10]
```diff
@@ -1,10 +1,49 @@
...
+        response
+    } = body;
+
+    try {
+        const result = framingFormat === HDLC
+            ? {frames: decodeFrames(body, 7).map(frame => decodeFrame(frame, body))}
```
Comment (DarkPark) [inline src/controllers/decoders/obisObserver.js]: would be nice to have some constant for 7 and 8

File: src/controllers/decoders/obisObserver.js [hunk 11]
```diff
@@ -1,10 +1,49 @@
...
+            : {commands: decodeMessage(bytes, body)};
+
+        reply.send({
+            protocol: 'obisObserver',
```
Comment (DarkPark) [inline src/controllers/decoders/obisObserver.js]: constants for protocols are also welcomed

File: tests/analogDecoderTest.js [hunk 12]
```diff
@@ -0,0 +1,182 @@
+import fastify from '../src/index.js';
```
Comment (DarkPark) [inline tests/analogDecoderTest.js]: do we need word `Test` in `analogDecoderTest.js`? each file here is a test

File: tests/analogDecoderTest.js [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+                    }]
+                }
+            }]
+        }
+    }
+];
+
+
+const testUrl = async ( url, request = {} ) => {
+    for ( let index = 0; index < tests.length; index++ ) {
```
Comment (DarkPark) [inline tests/analogDecoderTest.js]: it possible to simplify a little here with `tests.forEach`

Comment (DarkPark) [reply tests/analogDecoderTest.js]: also create individual `it` for each `tests` object

File: tests/analogSegmentsDecoder1Test.js [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+                    data: '23101007000000',
+                    assembledData: '23101007000000'
+                }]
+            }
+        }
+    }
+];
+
+
+const testUrl = async ( url, request = {} ) => {
```
Comment (DarkPark) [inline tests/analogSegmentsDecoder1Test.js]: can we move this function to utils?

File: tests/obisObserverEncoderTest.js [hunk 15]
```diff
@@ -0,0 +1,166 @@
+import fastify from '../src/index.js';
+import post from './utils/postRequest.js';
```
Comment (DarkPark) [inline tests/obisObserverEncoderTest.js]: `import postRequest from './utils/postRequest.js';`

or

`import post from './utils/post.js';`

File: tests/utils/postRequest.js [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+const serverUrl = `http://${configHTTP.host}:${configHTTP.port}`;
+
+const post = async ( url, request = {} ) => {
+    try {
+        return await axios.post(
+            `${serverUrl}${url}`,
+            request,
+            {
+                headers: {
+                    'Content-Type': 'application/json'
```
Comment (DarkPark) [inline tests/utils/postRequest.js]: there is a way to do it for each request - https://axios-http.com/docs/config_defaults

**Review by DarkPark — CHANGES_REQUESTED** [review]

File: package.json (hunk 2 above)
Comment (Pnom) [reply package.json]: fixed

File: src/routes/v1/utils/validateMtxRequest.js (hunk 5 above)
Comment (Pnom) [reply src/routes/v1/utils/validateMtxRequest.js]: fixed

File: tests/analogDecoderTest.js (hunk 13 above)
Comment (Pnom) [reply tests/analogDecoderTest.js]: fixed

File: tests/analogSegmentsDecoder1Test.js (hunk 14 above)
Comment (Pnom) [reply tests/analogSegmentsDecoder1Test.js]: fixed

File: tests/obisObserverEncoderTest.js (hunk 15 above)
Comment (Pnom) [reply tests/obisObserverEncoderTest.js]: fixed

File: tests/utils/postRequest.js (hunk 16 above)
Comment (Pnom) [reply tests/utils/postRequest.js]: fixed

File: src/routes/v1/utils/validateDecoderRequest.js (hunk 1 above)
Comment (Pnom) [reply src/routes/v1/utils/validateDecoderRequest.js]: fixed

File: src/routes/v1/utils/validateGeneralRequest.js (hunk 3 above)
Comment (Pnom) [reply src/routes/v1/utils/validateGeneralRequest.js]: fixed

File: tests/utils/postRequest.js [hunk 17]
```diff
@@ -0,0 +1,20 @@
+import axios from 'axios';
+import {http as configHTTP} from '../../src/configs/main.js';
+
+
+axios.defaults.baseURL = `http://${configHTTP.host}:${configHTTP.port}`;
+axios.defaults.headers.post['Content-Type'] = 'application/json';
+
+const post = async ( url, request = {} ) => {
+    try {
+        return await axios.post(
```
Comment (DarkPark) [inline tests/utils/postRequest.js]: better in a single line

File: tests/obisObserverEncoderTest.js [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+            bytesConversionFormat: BASE64,
+            framingFormat: HDLC,
+            frame: {
+                data: 'fgUBArvFfg=='
+            }
+        }
+    }
+];
+
+const routs = [
```
Comment (DarkPark) [inline tests/obisObserverEncoderTest.js]:
```suggestion
const routes = [
```
?

File: tests/utils/runTestsSequence.js [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+    describe(route.url, async () => tests.forEach( test => runTest(route, test)))
+);
+
+export const runTestsSequence = ( name, routs, tests ) => {
+    before(async () => {
+        await fastify.ready();
+    });
+
+    describe(name, async () => {
+        routs.forEach( route => runRouteTest(route, tests));
```
Comment (DarkPark) [inline tests/utils/runTestsSequence.js]:
```suggestion
        routs.forEach(route => runRouteTest(route, tests));
```

File: tests/obisObserverEncoderTest.js (hunk 18 above)
Comment (Pnom) [reply tests/obisObserverEncoderTest.js]: fixed

File: tests/utils/runTestsSequence.js (hunk 19 above)
Comment (Pnom) [reply tests/utils/runTestsSequence.js]: fixed

File: tests/utils/postRequest.js (hunk 17 above)
Comment (Pnom) [reply tests/utils/postRequest.js]: fixed

File: src/controllers/decoders/obisObserver.js (hunk 9 above)
Comment (Pnom) [reply src/controllers/decoders/obisObserver.js]: framingFormat is ok

File: src/controllers/decoders/mtxLora.js (hunk 8 above)
Comment (Pnom) [reply src/controllers/decoders/mtxLora.js]: fixed

File: src/controllers/utils/prepareCommands.js (hunk 6 above)
Comment (Pnom) [reply src/controllers/utils/prepareCommands.js]: fixed

File: src/controllers/utils/prepareCommands.js (hunk 7 above)
Comment (Pnom) [reply src/controllers/utils/prepareCommands.js]: fixed

File: src/controllers/decoders/obisObserver.js (hunk 11 above)
Comment (Pnom) [reply src/controllers/decoders/obisObserver.js]: fixed

File: src/controllers/decoders/obisObserver.js (hunk 10 above)
Comment (Pnom) [reply src/controllers/decoders/obisObserver.js]: fixed

File: src/constants/protocols.js [hunk 20]
```diff
@@ -0,0 +1,4 @@
+export const analog = 'analog';
```
Comment (DarkPark) [inline src/constants/protocols.js]:
```suggestion
export const ANALOG = 'analog'; 
```
?

Comment (Pnom) [reply src/constants/protocols.js]: fixed

**Pull request closed without merging**